### PR TITLE
Use component method calls for some of the envelope->weavelet calls.

### PR DESCRIPTION
**Pull request by ghemawat**
Previously, all envelope initiated communicaton to the weavelet was carried over a pair of pipes. This change switches some of that communication to be based on component method calls. In particular, the signal to start components is now carried by a method call (UpdateComponents) to a component hosted by the weavelet.

Main changes
------------

1. Introduce a controller component that is hosted by every remoteweavelet.
2. The envelope picks the name of a Unix domain socket and passes it to the weavelet it creates.
3. The envelope creates a stub pointing at this socket and makes calls to it to control the weavelet. (Currently only UpdateComponents is handled this way.)
4. remoteweavelet creates a controller implementation to handle method calls received over this Unix domain socket.

Other changes
-------------

1. envelope.NewEnvelope() now takes an Options struct as an argument. This struct can contain an optional Logger and an optional Tracer. The supplied Logger and Tracer are used for calls made to the controller component.
2. envelope.NewEnvelope() creates a temporary directory and allocates a Unix domain socket inside the directory. It also arranges to remove the directory when cleaning up (by catching termination signals for examples).
3. The socket name is passed in the EnvelopeInfo proto sent to the weavelet.
4. RemoteWeavelet creates a controller component and listens for calls to it on the supplied socket.
5. Added deployers.NewUnixSocketPath(), which allocates a new socket name inside a supplied directory.
6. weavertest multi deployer manages its own socket creation when it bypasses envelope.NewEnvelope().
7. website/blog/deployers/deployers_test.go kills child processes using SIGTERM instead of SIGKILL. This allows the child process a chance to clean up its temporary directories.
8. deployers_test.go logs the stderr and stdout lines it receives instead of dropping them on the floor. This can be very helpful when things aren't working as expected.
9. The deployer examples in the blog import the weaver package so they have access to the controller component registration.

### CODE REVIEW

File: runtime/envelope/envelope.go [hunk 1]
```diff
@@ -22,15 +22,22 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"log/slog"
 	"os"
 	"strconv"
 	"sync"
```
Comment (mwhittaker) [inline runtime/envelope/envelope.go:29]: What do you think about adding the following line right here? 

```go
_ "github.com/ServiceWeaver/weaver"
```

I believe this way, deployers don't have to import `"github.com/ServiceWeaver/weaver"` themselves.

File: controller.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+	"github.com/ServiceWeaver/weaver/runtime/protos"
+)
+
+// controller is a component hosted in every weavelet. Deployers make calls to this component to
+// fetch information about the weavelet, and to make it do various things.
+//
+// Arguments and results are protobufs to allow deployers to evolve independently of application
+// binaries.
+type controller interface {
```
Comment (mwhittaker) [inline controller.go:29]: Right now, the `controller` interface is duplicated here and in `envelope.go`. We could do the following to remove some of the duplication. 

First, create an `internal/control` package with a `controller.go` file:

```go
package control

import (
        "context"

        "github.com/ServiceWeaver/weaver/runtime/protos"
)

type Controller interface {
        GetHealth(context.Context, *protos.GetHealthRequest) (*protos.GetHealthReply, error)
        UpdateComponents(context.Context, *protos.UpdateComponentsRequest) (*protos.UpdateComponentsReply, error)
}
```

Second, update the `controller` definition here to the following:

```go
type controller control.Controller
```

Third, delete the `controller` interface in `envelope.go`. Additionally, replace all references to `controller` in `envelope.go` with `control.Controller`. 

Fourth, add the following line to `remoteweavelet.go`:

```go
var _ control.Controller = (*RemoteWeavelet)(nil)
```

What do you think? It's all a bit convoluted, but it does prevent us from changing the `controller` interface and forgetting to update either `RemoteWeavelet` or `Envelope`.

**Review by mwhittaker — APPROVED** [review]
Awesome change, Sanjay! When I tried to imagine what it would like to transition from using the pipe to using component method calls, I didn't foresee the change being so elegant. 

And very helpful CL description too, thanks! :pray:

File: runtime/envelope/envelope.go (hunk 1 above)
Comment (ghemawat) [reply runtime/envelope/envelope.go:29]: Good idea. Done.

File: runtime/envelope/envelope.go [hunk 3]
```diff
@@ -341,3 +398,38 @@ func dropNewline(line []byte) []byte {
...
+func getController(ctx context.Context, socket string, options Options) (controller, error) {
+	const controllerName = "github.com/ServiceWeaver/weaver/controller"
+	controllerReg, ok := codegen.Find(controllerName)
+	if !ok {
+		return nil, fmt.Errorf("controller component (%s) not found", controllerName)
+	}
+	controlEndpoint, err := call.ParseNetEndpoint("unix://" + socket)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (ghemawat) [reply runtime/envelope/envelope.go]: Done

File: runtime/envelope/envelope.go [hunk 4]
```diff
@@ -184,6 +231,9 @@ func NewEnvelope(ctx context.Context, wlet *protos.EnvelopeInfo, config *protos.
 	e.conn = conn
 	e.stdoutPipe = outpipe
 	e.stderrPipe = errpipe
+
+	removeDir = false  // Serve() is now responsible for deletion
+	cancel = func() {} // Delay real context cancellation
```
[the remark this answers is not in the document]
Comment (ghemawat) [reply runtime/envelope/envelope.go:239]: Thanks. There seems something wrong here (I don't think we are supposed to stash context values in structs, but I don't know how to fix it easily in the middle of this change.)

File: internal/weaver/remoteweavelet.go [hunk 5]
```diff
@@ -448,8 +462,13 @@ func (w *RemoteWeavelet) GetLoad(*protos.GetLoadRequest) (*protos.GetLoadReply,
...
 }
 
-// UpdateComponents implements the conn.WeaverHandler interface.
-func (w *RemoteWeavelet) UpdateComponents(req *protos.UpdateComponentsRequest) (*protos.UpdateComponentsReply, error) {
+// GetHealth implements the controls.Control interface.
+func (w *RemoteWeavelet) GetHealth(ctx context.Context, req *protos.GetHealthRequest) (*protos.GetHealthReply, error) {
+	return &protos.GetHealthReply{Status: protos.HealthStatus_HEALTHY}, nil
+}
+
+// UpdateComponents implements controls.Control and conn.WeaverHandler interfaces.
```
[the remark this answers is not in the document]
Comment (ghemawat) [reply internal/weaver/remoteweavelet.go]: Done

File: internal/weaver/remoteweavelet.go [hunk 6]
```diff
@@ -448,8 +462,13 @@ func (w *RemoteWeavelet) GetLoad(*protos.GetLoadRequest) (*protos.GetLoadReply,
 	return &protos.GetLoadReply{Load: report}, nil
 }
 
-// UpdateComponents implements the conn.WeaverHandler interface.
-func (w *RemoteWeavelet) UpdateComponents(req *protos.UpdateComponentsRequest) (*protos.UpdateComponentsReply, error) {
+// GetHealth implements the controls.Control interface.
```
[the remark this answers is not in the document]
Comment (ghemawat) [reply internal/weaver/remoteweavelet.go]: Done

File: internal/weaver/remoteweavelet.go [hunk 7]
```diff
@@ -55,6 +56,8 @@ type RemoteWeaveletOptions struct {
 // coordinates with a deployer over a set of Unix pipes to start other
 // components remotely. It is the weavelet used by all deployers, except for
 // the single process deployer.
+//
+// RemoteWeavelet must implement the weaver.Controls component interface.
```
[the remark this answers is not in the document]
Comment (ghemawat) [reply internal/weaver/remoteweavelet.go]: Done

File: controller.go (hunk 2 above)
Comment (ghemawat) [reply controller.go]: Great idea! Done.

**Review by ghemawat — COMMENTED** [review]
Thanks!

**Pull request merged**